Support for multi-container deployments and bootcamp improvements

antoniopetrole opened the pull request:
## Overview
I took the time to implement support for multi-container deployments using Docker compose. As someone who has worked with Greenplum for 9 years, the reason for this is that a single node instance of Cloudberry significantly inhibits a users ability to understand and learn about the distributed nature of Cloudberry. While implementing this work, I found a number of issues, bugs, typos, and slightly confusing statements in the tutorials which I went through multiple times. While I discovered these issues, I spent the time to correct them along with the multi-container work. I'm more than happy to discuss with anyone why certain changes were made as I made sure I was thoughtful when making any changes to the existing code base. I'm hoping these changes serve the Cloudberry community moving forward

**New Features**
1. Implemented multi-container docker compose functionality. This functionality will allow future lessons around things like high availability for example and demonstrates the distributed nature of Cloudberry. 
- Added new compose yml files
- Modified init_system scripts to conditionally run between multimode or single node with MULTINODE env variable
- Implemented multi-node deployment in run.sh script with -m flag
- Modified references to localhost in commands to now reference mdw so it works on both single and multi container deployments
- Modified deployed container names so docker exec command is more straightforward between deployment options
- Included a diagram with descriptions highlighting the difference between the single container and multi-container deployment options
- Updated gpbackup lesson to use gpssh to see backup files. This makes the command portable between the single container and multi-container deployment options 
- Removed documentation references to "Single-Node" deployments where now inappropriate and replaced diction with "Docker Based" deployments.

**New Improvements and Bug Fixes**
1. Added smoke tests for tutorial up to lesson 5. This file can be found in /tmp/smoke-test.sh and should be run as gpadmin. This test runs all of the commands in lessons 1-5 to validate we're not making breaking changes during development. It's a fairly naive implementation since it's in bash but we should be able to use it for other things moving forward. Tutorial 6 requires you to build gpbackup and I was concerned about compatibility issues around the idea of building gpbackup into the dockerfile.
2. Docker exec command in 000 tutorial now references the cbdb-mdw container name. Previously if you deployed any containers **after** deploying CBDB and ran the previous exec command, you would exec into the last deployed container and not the cbdb coordinator.
3. For section 101-5-queries and performance tuning, the example to create table faa.otp_c already exists when we created and loaded it in 101-4-data-loading.md. This is now fixed by dropping the existing table before recreating it
4. Query in “Indexes and Performance” doesn’t have schema name prefixed for the table. This makes the assumption gpadmin has the search_path for schema set to faa which it does not. Added schema name prefixes and now it's fine
5. Added section to 000 lesson highlighting useful commands for starting, stopping, and cleaning up the environment. 
6. Fixed various typos in lesson .md files

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cloudberrydb/bootcamp?pullRequest=33) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cloudberrydb/bootcamp?pullRequest=33) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cloudberrydb/bootcamp?pullRequest=33) it.</sub>

edespino reviewed (commented):
### Code Review: Validation of Docker Compose in Container-Optimized OS (COS) VM on GCP

#### Context
To help validate this PR, I have spun up a Container-Optimized OS (COS) VM in Google Cloud Platform (GCP). The default environment doesn't have the docker compose command. I know ... why does Eddie have to be different (I digress).

#### Steps Taken
Essentially, I ran through the instructions here to bring in Docker Compose:

##### Scenario two: Install the Compose plugin
- Followed the steps to install the Docker Compose plugin.

#### Observations
- I like the note indicating that Docker Compose is a requirement.
- However, if the user provides the run.sh -m option, a validation for the existence of the Docker Compose command is not performed.
- The error output below is received and the docker multi-node is not started.

#### Recommendations
- Add a validation check at the beginning of the script to verify if Docker Compose is installed when the -m option is provided.
- If Docker Compose is not found, the script should exit gracefully and provide an informative error message to the user.

#### Example Error Message
If Docker Compose is not installed, the script will fail with an error message similar to the following:

```sh
Error: Docker Compose command not found. Please install Docker Compose by following the instructions at https://docs.docker.com/compose/install/#scenario-two-install-the-compose-plugin.
```

#### Conclusion
Including this validation will ensure that the script does not proceed with the assumption that Docker Compose is available, thus preventing runtime errors and improving the user experience.

#### Error output
```
Step 20/20 : CMD ["bash","-c","/tmp/init_system.sh"]
 ---> Running in 7b6581aa52e7
Removing intermediate container 7b6581aa52e7
 ---> c5f375ef2b2f
Successfully built c5f375ef2b2f
Successfully tagged cbdb-1.5.4:centos7
unknown shorthand flag: 'f' in -f
See 'docker --help'.

Usage:  docker [OPTIONS] COMMAND

A self-sufficient runtime for containers

Common Commands:
  run         Create and run a new container from an image
  exec        Execute a command in a running container
  ps          List containers
  build       Build an image from a Dockerfile
  pull        Download an image from a registry
  push        Upload an image to a registry
  images      List images
  login       Log in to a registry
  logout      Log out from a registry
  search      Search Docker Hub for images
  version     Show the Docker version information
  info        Display system-wide information
```

edespino commented on 000-cbdb-sandbox/README.md:
````diff
@@ -102,5 +122,59 @@ In addition to using the `docker exec` command, you can also use the `ssh` comma
+docker compose -f <docker-compose-OS.yml> down
+```
+
+**Starting A Stopped Cloudberry Docker Deployment**
````
Maybe be explicit and introduce single and multi container sub-headings.

antoniopetrole replied:
Added in commit https://github.com/cloudberrydb/bootcamp/pull/33/commits/1bf9949713e086b19cb827542824295ee565b50f

edespino commented on 000-cbdb-sandbox/configs/smoke-test.sh:
```diff
@@ -0,0 +1,216 @@
+set -e
```
The script is missing a shebang at the top (#!/bin/bash) for proper interpretation by the shell.

edespino replied:
Additionally, please ensure [ShellCheck](https://github.com/koalaman/shellcheck) has been run on the scripts to catch and address any obvious errors. Focus on addressing issues related to the changes you introduced.

edespino replied:
BTW: I'm really digging that you introduced this file. GREAT WORK!

antoniopetrole replied:
Added in the commit below. I've never used ShellCheck but really digging it, never considered using a linter for bash in the past for some reason
https://github.com/antoniopetrole/bootcamp/commit/6be8307b3c8f8eedd0de89cf7e34cc6ef74fb36e

edespino commented on 000-cbdb-sandbox/configs/smoke-test.sh:
```diff
@@ -0,0 +1,216 @@
+    echo "################################"
+
+    cd /tmp
+    tar xzf faa.tar.gz
```
In my environment, the file /tmp/faa.tar.gz was not readable by gpadmin and thus the tar failed.

```
-rw-r----- 1 root    root    63088802 Aug  6 08:25 faa.tar.gz
```

antoniopetrole replied:
Strange, it's owned by root for me as well however as gpadmin I can still untar it. Changed the init_system script to make gpadmin the owner just to be safe
https://github.com/antoniopetrole/bootcamp/commit/51c5727a455b85f60c2697ee504f6d11b2b645ea

edespino commented on 000-cbdb-sandbox/README.md:
````diff
@@ -64,7 +84,7 @@ You can now connect to the database and try some basic operations.
 
     ```shell
-    docker exec -it $(docker ps -q) /bin/bash
+    docker exec -it cbdb-mdw /bin/bash
````
With the introduction of the multi-node container configuration, it may take extra time for the cluster to boot. Consider adding a note to inform users that the gpinitsystem process may take additional time to complete.

edespino replied:
A thought occurred (especially for the multi-node config), to include a `select * from gp_segment_configuration` here to demonstrate the multi-node config in all it's cloudberry glory. Just a thought.

antoniopetrole replied:
For the gp_segment_configuration comment, I made sure to add a note to check the docker logs to see if your database is finished initializing. The gp_segment_configuration output can be found there (you shouldn't need to scroll either). I figured it's best to keep it there than in the README so the README doesn't get too verbose (like should we include output for both single vs multi container deployments to not confuse end users, this was a constant battle I was dealing with when implementing that feature).
https://github.com/antoniopetrole/bootcamp/commit/1bf9949713e086b19cb827542824295ee565b50f

Also added a note to the same commit about how the cluster may take a few minutes to start up (it's really only about 30 seconds on my machine but wanted to be liberal about that timing in case people are using older machines).

edespino commented on 000-cbdb-sandbox/README.md:
````diff
@@ -102,5 +122,59 @@ In addition to using the `docker exec` command, you can also use the `ssh` comma
+To stop the **single container** deployment and also remove the volume that belongs to the container, you can run the following command. Keep in mind this will remove the volume as well as the container associated which means any changes you've made inside of the container or any database state will be wiped and unrecoverable.
+
+```shell 
+docker rm -f cbdb-mdw
````
I noticed there is no corresponding command for the multi container config. I actually ran this and noticed I had three containers remaining :)

antoniopetrole replied:
I think this was confusion since there weren't the subheadings from your afforementioned comments on this PR. Now that those subheadings are there it should hopefully stop users from running this command and scrolling down to the "Multi-container" deployment subheading. Might make more sense though to split the useful commands section by deployment types instead of user intentions like how it is now (see this comment as well https://github.com/cloudberrydb/bootcamp/pull/33#discussion_r1707659964)

edespino reviewed (changes requested):
Please see my comments scattered about. Very nice work thus far. Keep up the great work. I ran through all combinations of configurations including those you have introduced.

antoniopetrole commented on 000-cbdb-sandbox/README.md:
```diff
@@ -102,5 +122,59 @@ In addition to using the `docker exec` command, you can also use the `ssh` comma
+
+When working with the Cloudberry Docker environment there are a few commands that will be useful to you.
+
+**Stopping Your Single Container Deployment With Docker**
```
Great feedback, added in the following commit https://github.com/cloudberrydb/bootcamp/pull/33/commits/1bf9949713e086b19cb827542824295ee565b50f

antoniopetrole replied:
I could also change this to just make a big section for "Useful single-container deployment commands" and then a big section for "Useful multi-container deployment commands". 6 of one half dozen of the other I suppose

antoniopetrole commented:
Also just want to add a quick comment that I forgot to mention in my original PR, there is instructions where you can SSH to the container instead of using `docker exec`. I don't think this is a super wise idea since it relies on port forwarding port 22 on the local host to 22 on the coordinator container. Reason I think this might cause issues is that if people are running ssh tunnels/forwarding on their local machine, then trying to start the container will result in an error with that port already being used. This may be an edge case that's not worth discussing and I'm open to keeping it how it is, I'll just need to add that port forwarding do Docker compose since it's currently not implemented

tuhaihe reviewed (commented):
Thanks @antoniopetrole for your great work!

tuhaihe commented:
Hi @antoniopetrole, I tested all four sandbox modes, and they all built successfully. However, I received some warnings:

1. When I ran `./run.sh -m`, I got the following warning:

```
WARN[0000] /root/bootcamp/000-cbdb-sandbox/docker-compose-centos7.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion 
```

Then I ran the command `docker exec -it cbdb-mdw /bin/bash` but failed: 

```
Error response from daemon: No such container: cbdb-mdw
```

2. when run `./run.sh -c main -o rockylinux9 -m`, still got warning:

```
 WARN[0000] /root/bootcamp/000-cbdb-sandbox/docker-compose-rockylinux9.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion  
```

Thanks again!

antoniopetrole commented on 000-cbdb-sandbox/README.md:
```diff
@@ -19,15 +19,23 @@ Make sure that your environment meets the following requirements:
 ## Build the Sandbox
 
-This section introduces two methods to set up the Docker container. The container will host a CBDB single-node cluster intialized with one coordinator and three primary and mirror segments. Both x86 and ARM CPUs (including Apple chips) are supported.
+When building and deploying CloudBerry in Docker, you will have 2 different deployment options as well as different build options.
```
Good catch and fixed below

https://github.com/cloudberrydb/bootcamp/pull/33/commits/0e18cc5e034eac0785a12ca8faf20ddf0ddc07fe

antoniopetrole commented:
> Hi @antoniopetrole, I tested all four sandbox modes, and they all built successfully. However, I received some warnings:
> 
> 1. When I ran `./run.sh -m`, I got the following warning:
> 
> ```
> WARN[0000] /root/bootcamp/000-cbdb-sandbox/docker-compose-centos7.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion 
> ```
> 
> It was built successfully. When I ran the command `docker exec -it cbdb-mdw /bin/bash` but failed:
> 
> ```
> Error response from daemon: No such container: cbdb-mdw
> ```
> 
> 2. when run `./run.sh -c main -o rockylinux9 -m`, still got warning:
> 
> ```
>  WARN[0000] /root/bootcamp/000-cbdb-sandbox/docker-compose-rockylinux9.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion  
> ```
> 
> Thanks again!

Interestingly enough even after upgrading Docker Desktop to the latest version I still don't get this warning, however I removed it since it's deprecated anyways. Also the reason you received the "No such container" error must have been from the CentOS docker compose file. I had the intention of removing CentOS as an option based on the feedback from the slack channel but wanted to not go too crazy with scope creep on this PR. So when I added the "container-name" attributes to the rocky compose file I didn't add it to the CentOS one. That should all be fixed now
https://github.com/cloudberrydb/bootcamp/pull/33/commits/ba0a7b1ebf3e92a3fdf141c3822882d7b2a2fe49

antoniopetrole commented:
> ### Code Review: Validation of Docker Compose in Container-Optimized OS (COS) VM on GCP
> #### Context
> To help validate this PR, I have spun up a Container-Optimized OS (COS) VM in Google Cloud Platform (GCP). The default environment doesn't have the docker compose command. I know ... why does Eddie have to be different (I digress).
> 
> #### Steps Taken
> Essentially, I ran through the instructions here to bring in Docker Compose:
> 
> ##### Scenario two: Install the Compose plugin
> * Followed the steps to install the Docker Compose plugin.
> 
> #### Observations
> * I like the note indicating that Docker Compose is a requirement.
> * However, if the user provides the run.sh -m option, a validation for the existence of the Docker Compose command is not performed.
> * The error output below is received and the docker multi-node is not started.
> 
> #### Recommendations
> * Add a validation check at the beginning of the script to verify if Docker Compose is installed when the -m option is provided.
> * If Docker Compose is not found, the script should exit gracefully and provide an informative error message to the user.
> 
> #### Example Error Message
> If Docker Compose is not installed, the script will fail with an error message similar to the following:
> 
> ```shell
> Error: Docker Compose command not found. Please install Docker Compose by following the instructions at https://docs.docker.com/compose/install/#scenario-two-install-the-compose-plugin.
> ```
> 
> #### Conclusion
> Including this validation will ensure that the script does not proceed with the assumption that Docker Compose is available, thus preventing runtime errors and improving the user experience.
> 
> #### Error output
> ```
> Step 20/20 : CMD ["bash","-c","/tmp/init_system.sh"]
>  ---> Running in 7b6581aa52e7
> Removing intermediate container 7b6581aa52e7
>  ---> c5f375ef2b2f
> Successfully built c5f375ef2b2f
> Successfully tagged cbdb-1.5.4:centos7
> unknown shorthand flag: 'f' in -f
> See 'docker --help'.
> 
> Usage:  docker [OPTIONS] COMMAND
> 
> A self-sufficient runtime for containers
> 
> Common Commands:
>   run         Create and run a new container from an image
>   exec        Execute a command in a running container
>   ps          List containers
>   build       Build an image from a Dockerfile
>   pull        Download an image from a registry
>   push        Upload an image to a registry
>   images      List images
>   login       Log in to a registry
>   logout      Log out from a registry
>   search      Search Docker Hub for images
>   version     Show the Docker version information
>   info        Display system-wide information
> ```

This should be fixed now I hope, it's a bit challenging to recreate this issue in my environment but this check will catch any non-zero status codes which based on our slack discussions is happening to you here (as well as when I tested it in my RHEL 9 image locally)
https://github.com/cloudberrydb/bootcamp/pull/33/commits/fab6c97d88c456ca615d4263187e671242c0a299

antoniopetrole commented:
@edespino @tuhaihe As far as I can tell all of the feedback has been implemented into the code base, thanks again for both of you testing these changes as it's not easy to replicate some of the weird docker things you run into during development by yourself. Happy to make any other changes, wasn't sure if I should be resolving comments when I update them with commits or not so I left them open as I figured whoever is requesting the comment should determine if the success criteria was met.

antoniopetrole commented:
> Also just want to add a quick comment that I forgot to mention in my original PR, there is instructions where you can SSH to the container instead of using `docker exec`. I don't think this is a super wise idea since it relies on port forwarding port 22 on the local host to 22 on the coordinator container. Reason I think this might cause issues is that if people are running ssh tunnels/forwarding on their local machine, then trying to start the container will result in an error with that port already being used. This may be an edge case that's not worth discussing and I'm open to keeping it how it is, I'll just need to add that port forwarding do Docker compose since it's currently not implemented

Just want to follow up on this comment btw. I tested the regular single container versions and the ssh command doesn't work on them either (even before my commits). I looks like the docker run command is actually forwarding locally on port 122 not 22, so the ssh command has no way of working unless you pass that in. I set just as a test (but didn't commit it) forwarding 22:22 on the coordinator container which does work for ssh, just wanted to make sure someone saw this feedback as the documentation currently reflects a broken command as far as I can tell

edespino commented on 000-cbdb-sandbox/README.md:
````diff
@@ -102,5 +122,63 @@ In addition to using the `docker exec` command, you can also use the `ssh` comma
+docker compose -f <docker-compose-OS.yml> stop
+```
+
+To stop the **multi-container** deployment while _keeping the data and state_ within the container, you can run the command below by subsituting the <docker-compose-OS.yml> argument with whichever docker-compose file is associated with your deployment (Rocky Linux 9 or CentOS). Running this command means it will delete the containers as well as remove the volumes that the containers are associate ...
+
+```shell 
+docker compose -f <docker-compose-OS.yml> down
+```
````
I believe the initial sentence is incorrect as the `down` operation will delete the containers. The sentence indicates the data and state are kept.

antoniopetrole replied:
You're correct on this, looks like it was maybe a copy-paste that didn't get corrected. Fixed below, also added the -v flag which I started using today since otherwise it'll leave you with dangling volumes that can't be reattached later if you use down
https://github.com/antoniopetrole/bootcamp/commit/a8fb88a41f3de46245e5860d086a8362e5d3061f

edespino reviewed (changes requested):
One more change.

edespino reviewed (approved):
Thank you for your contribution. I am approving this PR. You can squash your commits and merge. Please let me know if you need any assistance with the merge. Cheers, -=e
